mpa/cls/stage.py

```diff
@@ -206,20 +202,23 @@ def configure_task(cfg, training, model_meta=None, **kwargs):
                 if not cfg.model.get('multilabel', False):
                     efficient_mode = cfg['task_adapt'].get('efficient_mode', True)
                     gamma = 2 if efficient_mode else 3
-
-                    cfg.model.head.loss = ConfigDict(
-                        type='SoftmaxFocalLoss',
-                        loss_weight=1.0,
-                        gamma=gamma,
-                        reduction='none',
-                    )
                     sampler_type = 'balanced'
+                    if len(model_classes) > 0:
+                        cfg.model.head.loss = ConfigDict(
+                            type='SoftmaxFocalLoss',
+                            loss_weight=1.0,
+                            gamma=gamma,
+                            reduction='none',
+                        )
+                    elif len(model_classes) == 0:
```

Comment:
@supersoob I think it's best to just use 'else'.

Reply:
Also, I don't think model_classes will always be zero for the Supervised Learning setting, what do you think?
For active learning and supervised learning, model_classes is expected to be greater than zero.

Reply:
After reading your comment, I thought about the case that the supervised should be run. I'd say that when model classes and data classes are totally different without intersection, supervised learning need to be initiated. Otherwise, should start incremental learning.
So I plan to change this part
`if len(set(model_classes) & set(dst_classes))==0: #change loss`
what do you think? have any better idea?
And I will change the default op into "REPLACE" to get the dst_classes correctly as you discussed in your previous PR.

Reply:
I've been thinking about this for a little bit..
For supervised learning, I think there are two cases:
- where model_classes and data_classes are the same
- or model_classes = []

```
if len(model_classes) == 0 or model_classes == data_classes:
  cfg.model.head.loss = dict(type='CrossEntropyLoss', loss_weight=1.0)
else:
  cfg.model.head.loss = ConfigDict(
      type='SoftmaxFocalLoss',
      loss_weight=1.0,
      gamma=gamma,
      reduction='none',
  )
```
In this case, I think you might have to look at the order of model_classes and data_classes (model_classes == data_classes).

There may be some parts that I missed, so please let me know if there are any.

Reply:
Agree that these two cases are for supervised learning.

- where model_classes and data_classes are the same -> No new classes

- or model_classes = [] -> model never been trained

And in my opinion, I think the case I mentioned in the last comment should be considered, too.

- len(set(model_classes) & set(dst_classes))==0 -> for example, the model has trained with some or large-scale data, but want to optimize it to my custom data

For example, there is a model trained with label [a,b], but the label schema from data is [c,d].
And they are totally different and no intersection here.
In that case, I'd say this is not incremental and the model is to run in the way of supervised learning/finetune.
Plus, since we already consented to enable REPLACE mode, the dst_classes will be replaced by [c,d] and the newly trained model's label schema will be turned to [c,d]. If this case is deemed as incremental, MERGE op should be set.

I don't know the correct answer about dealing the learning type, since there are so many cases and perspectives.
If you have any other opinion, please comment freely:)

Reply:
- model_classes = [] -> SL from scratch **-> Yes**
- model_classes is not empty and label_schema changes -> C-IL **-> Yes, but may have a problematic case like you mentioned (REPLACE + Incremental) **
- but, [a, b] -> [c, d] case means you need to learn like SL, right? **-> Yes**

To clearly explain the third case,

-  model class = [a,b]
-  data class = [c,d]

Want to deal as incremental?
dst_classes = [a,b,c,d] (MERGE op required in current code base)

Want to deal as supervised?
dst_classes = [c,d] (REPLACE op required in current code base)

I think this should be deemed supervised learning case. Do you agree with it?

However, I found a problematic case in both REPLACE and Incremental mode you mentioned.
When model classes and data classes have same classes partially, like
model class = [a,b]
data class = [b,c]
dst_classes = [b,c] (REPLACE op)
dst_classes = [a,b,c] (MERGE op)

final dst_classes will be same as data classes when REPLACE op is enabled. If it learns in incremental mode, it would not learn some original model classes, which are not included in data classes (and actually it is not in data folder). In SC, since the data folder path is constant, we don't need to worry if data class is not a superset of model_class (or maybe not when classes are removed from the user). But when pre-trained model is loaded manually and partially has same classes with dataset, I wonder which learning mode the user wants (run incrementally? or from scratch with dataset?). If the case is for supervised learning, we can leave REPLACE op as default and need additional 'if condition' to deal with this loss change code. If not, MERGE op is needed only for this case.

If I missed anything, please let me know.

Reply:
Then disenabling REPLACE and turning on MERGE is required for incremental learning in the problematic case I mentioned. I realized that it's orthogonal to the user's decision made by SC checkbox (whether scratch or not) in OTE code. Could you share your opinion about the third case or else?

Reply:
I changed the codes here reflecting this discussion. Thank you for giving meaningful comments:)